README.md
Unicode Lookup Table
====================

[![Build Status](https://github.com/detomon/unicode-table/actions/workflows/c.yml/badge.svg?branch=master)](https://github.com/detomon/unicode-table/actions/workflows/c.yml)

This script generates a Unicode character lookup table with linear access time. It creates a header and source file and compiles a static library usable within C/C++. The source data is contained in the files [UnicodeData.txt](https://www.unicode.org/Public/15.0.0/ucd/UnicodeData.txt) and [SpecialCasing.txt](https://www.unicode.org/Public/15.0.0/ucd/SpecialCasing.txt) and can be found on <https://www.unicode.org/Public/UNIDATA/>. Currently Unicode version 15.0.0 is used, but the files can be replaced with newer versions in the future.

The only function is `UTLookupGlyph` that looks up a single character by its Unicode value. It returns a pointer to a `UTInfo` struct containing the character informations. It always returns a valid pointer, even for invalid characters. In this case, the field `category` has the value `UT_CATEGORY_INVALID` assigned.

Available Informations
----------------------

`UTInfo` contains the following fields:

```c
typedef struct {
    uint32_t flags;    ///< Combination of UTFlag.
    uint32_t category; ///< One of UTCategory.
    int32_t cases[3];  ///< Distance to case variant. Indexable with UTCase.
    union {
        int64_t num;      ///< Number value if `flags & UT_FLAG_NUMBER`.
        char const* frac; ///< Fraction string if `flags & UT_FLAG_FRACTION`.
    };
} UTInfo;
```

- `category` contains one of the character categories listed in `UTCategory`
- `flags` contains multiple flags listed in `UTFlag`
- `cases` contains values to be added to the character value in order to convert it to the desired case variant (`UT_CASE_UPPER`,`UT_CASE_LOWER` or `UT_CASE_TITLE`). The field is indexable with `UTCase`. If a `case` field is `0`, that specific case variant does not exist or is the same case variant as the character value itself. If one of the flags `UT_FLAG_UPPER_EXPANDS`, `UT_FLAG_LOWER_EXPANDS` or `UT_FLAG_TITLE_EXPANDS` is set in `flags`, the character expands to multiple characters when case-folding. For example, the lowercase letter "ß" (`0x00DF; LATIN SMALL LETTER SHARP S `) expands to the 2 uppercase letters "SS" (`0x0053 0x0053; LATIN CAPITAL LETTER S `). `cases` then contains an index usable for the array `UTSpecialCases`. The index itself points to the number of character in the expanded sequence. The following array elements contain the expanded sequence's character values (see [example below](#user-content-case-fold-expansion)).
- `number` contains numeric values for digits, number-like and fraction characters. For example, the roman number "Ⅶ" (`0x2166; ROMAN NUMERAL SEVEN `) has the value `7` in `num`. Fractions are represented by strings that contain the nominator and denominator separated by `/` (`"n/d"`). For example, the fraction character "¼" (`0x00BC; VULGAR FRACTION ONE QUARTER `) has the value `"1/4"` in `frac`.

Lookup Character
----------------

Looking up a character with its Unicode value:

```c
// character `Đ` (0x0110; LATIN CAPITAL LETTER D WITH STROKE)
UTGlyph glyph = 0x0110;
UTInfo const* info = UTLookupGlyph(glyph);

// get lowercase variant `đ` (0x0111; LATIN SMALL LETTER D WITH STROKE)
UTGlyph lower = glyph + info->cases[UT_CASE_LOWER];

// prints "Lowercase variant of 0x0110: 0x0111"
printf("Lowercase variant of 0x%04X: 0x%04X\n", glyph, lower);
```

Numeric and Fraction Values
---------------------------

Get the representing integer or fraction value:

```c
UTGlyph glyph;
UTInfo const* info;

// character `Ⅶ` (0x2166; ROMAN NUMERAL SEVEN)
glyph = 0x2166;
info = UTLookupGlyph(glyph);

// check if character is a number
if (info->flags & UT_FLAG_NUMBER) {
    // prints "Integer value of 2166: 7"
    printf("Integer value of %04X: %lld\n", glyph, info->num);
}

// character `¼` (0x00BC; VULGAR FRACTION ONE QUARTER)
glyph = 0x00BC;
info = UTLookupGlyph(glyph);

// check if character is a fraction
if (info->flags & UT_FLAG_FRACTION) {
    // prints "String representation of 0x00BC: 1/4"
    printf("String representation of 0x%04X: %s\n", glyph, info->frac);
}
```

Case-Fold Expansion
-------------------

Handling cases, where case-folding expands to multiple characters:

```c
// character `ß` (0x00DF; LATIN SMALL LETTER SHARP S)
UTGlyph glyph = 0x00DF;
UTInfo const* info = UTLookupGlyph(glyph);

// check if expansion occurs to prevent invalid index
if (info->flags & UT_FLAG_UPPER_EXPANDS) {
    // sequence index for uppercase variant
    int idx = info->cases[UT_CASE_UPPER];
    int length = UTSpecialCases[idx];

    // character sequence
    UTGlyph const* sequence = &UTSpecialCases[idx + 1];

    // prints "0x00DF expands to 2 chars in uppercase"
    printf("0x%04X expands to %d chars in uppercase\n", glyph, length);

    // uppercase characters
    // prints:
    // "0: 0x0053"
    // "1: 0x0053"
    for (int i = 0; i < length; i ++) {
        printf("%d: 0x%04X\n", i, sequence[i]);
    }
}
else {
    printf("Character %04X does not expand\n", glyph);
}
```

Building the Table
------------------

Run `./autogen.sh` to generate the build system. Then run `./configure` and `make` to build `libunicodetable.a` in the `src` directory:

```sh
./autogen.sh
./configure
make
```

`configure` can take some options:

- `--enable-symbol-prefix=bla` changes the prefix for library symbols to `bla`. The default is `UT`.
- `--enable-snake-case`, `--disable-snake-case` enables or disables snake-case symbol names (For example, `bla_lookup_glyph` instead of `blaLookupGlyph`). The default is `disable`.
- `--enable-categories=Lu,Ll,Lt,Lm,Lo,Mn,Mc,Me,Nd,Nl,No,Pc,Pd,Ps,Pe,Pi,Pf,Po,Sm,Sc,Sk,So,Zs,Zl,Zp,Cc,Cf,Cs,Co,Cn` sets the required Unicode character categories to be include in the table. This can reduce the table size. All other characters will have their category set to `UT_CATEGORY_OTHER_NOT_ASSIGNED`. If omitted, all categories are included.
- `--enable-include-info=flags,categories,casing,numbers` sets the required character informations to be included in the table. If omitted, all available informations are included.
- `--enable-strict-level=0` sets the strict level, which excludes certain characters considered as unsafe, for example, surrogates. Defaut is 0.  
0: do not exclude any characters  
1: define surrogates as invalid  

### Example

```sh
./configure \  
    --enable-symbol-prefix=bla \  
    --enable-categories=Lu,Ll,Lt,Lm,Lo \  
    --enable-snake-case \  
    --enable-strict-level=1
```

Using as Automake Subproject
----------------------------

This project is designed to be used as an Automake subproject. To match your projects namespace, export the enviroment variables inside the main project's `configure.ac`:

```sh
...

# put before `AC_CONFIG_SUBDIRS`

# use symbol prefix
# omit for default prefix
export UT_SYMBOL_PREFIX="myprefix"
# use snake case symbols
export UT_SNAKE_CASE=1
# define character categories to include
# omit to include all categories
export UT_CATEGORIES=Lu,Ll,Lt
# define character information to include
# omit to include all available information
export UT_INCLUDE_INFO=flags,categories
# exclude additional characters, for example, surrogates
# omit for using default (0)
# 0: do not exclude any characters
# 1: define surrogates as invalid
export UT_STRICT_LEVEL=0

AC_CONFIG_SUBDIRS([unicode-table])

...
```

Source Templates
----------------

The source templates `unicode-table.h.in` and `unicode-table.c.in` located in `src` contain the structure and data placeholders for the header and source file. They can be editted if needed.


ChangeLog
unicode-table 0.5.5:
  Release date: 2023-05-06

* Fix UTGetSpecialCase

unicode-table 0.5.4:
  Release date: 2022-10-27

* Update Unicode data version to 15.0.0
* Update build config

unicode-table 0.5.3:
  Release date: 2020-09-14

* Cleanup generator script
* Fix version string

unicode-table 0.5.2:
  Release date: 2020-04-05

* Update Unicode data

unicode-table 0.5.1:
  Release date: 2018-04-02

* Update source templates
* Update README.md

unicode-table 0.5.0:
  Release date: 2017-07-24

* Update Unicode data version to 10.0.0

unicode-table 0.4.1:
  Release date: 2017-05-26

* Update header template
* Update generation script
* Added function to check UTF-8 glyph for overlong sequence
* Adel function to lookup special casing

unicode-table 0.4:
  Release date: 2017-02-25

* Add strict level option

unicode-table 0.3.3:
  Release date: 2016-09-19

* Update Unicode data version to 9.0.0

unicode-table 0.3.2:
  Release date: 2016-06-02

* Updated source templates

unicode-table 0.3.1:
  Release date: 2016-03-25

* Set optimal integer type to reduce table size

unicode-table 0.3:
  Release date: 2016-03-20

* Add category and info build options
* Fix reading character ranges

unicode-table 0.2:
  Release date: 2016-03-13

* Update symbol names
* Use uppercase constants
* Add better support as Automake subproject

unicode-table 0.1:
  Release date: 2015-11-17

* First release


test.c
#include <stdio.h>
#include <string.h>
#include "unicode-test.h"

static int getSequence (char const * sequence, int values [])
{
	return sscanf (sequence, " %x %x %x %x", &values [0], &values [1], &values [2], &values [3]);
}

int main (int argc, char const * argv [])
{
	if (argc <= 1) {
		fprintf (stderr, "usage: %s UnicodeData.txt SpecialCasing.txt\n", argv [0]);
		return 1;
	}

	char const * dataFile = argv [1];
	char const * specialCasingFile = argv [2];

	FILE * data = fopen (dataFile, "r");
	FILE * specialCasing = fopen (specialCasingFile, "r");

	char line [1024];
	unsigned value;
	char name [63];
	char categoryName [3];

	UTInfo const * info;
	UTCategory category;

	printf ("checking...\n");
	printf ("reading %s...\n", dataFile);

	// read UnicodeData.txt
	while (fgets (line, 1024, data)) {
		sscanf (line, "%x;%63[^;];%2[^;]", &value, name, categoryName);

		info = UTLookupRune (value);
		category = info -> category;

		// test category
		if (strcmp (UTCategoryNames [category], categoryName) != 0) {
			fprintf (stderr, "ERROR: %x: %s != %s (%d)\n", value, categoryName, UTCategoryNames [category], category);
			return 1;
		}
	}

	char upperSequence [64];
	char lowerSequence [64];
	char titleSequence [64];
	int length;
	int values [4];

	printf ("reading %s...\n", specialCasingFile);

	// read SpecialCasing.txt
	while (fgets (line, 1024, specialCasing)) {
		for (int i = 0; line [i]; i ++) {
			if (line [i] == '#') {
				line [i] = '\0';
				break;
			}
		}

		strcpy (name, "");

		if (sscanf (line, "%x ; %63[^;] ; %63[^;] ; %63[^;] ; %63[^;]", &value, lowerSequence, titleSequence, upperSequence, name) < 3) {
			continue;
		}

		// ignore conditional case-folding
		if (name [0]) {
			continue;
		}

		info = UTLookupRune (value);

		if (!(info -> flags & (UTFlagUpperExpands | UTFlagLowerExpands | UTFlagTitleExpands))) {
			fprintf (stderr, "No special case found for '%04x'\n", value);
			return 1;
		}

		length = getSequence (upperSequence, values);

		if (length > 1) {
			if (!(info -> flags & UTFlagUpperExpands)) {
				fprintf (stderr, "Missing upper special seqeunce for '%04x'", value);
				return 1;
			}

			if (UTSpecialCases [info -> cases [UTCaseUpper]] != length) {
				fprintf (stderr, "Wrong upper seqeunce length for '%04x'", value);
				return 1;
			}
		}

		length = getSequence (lowerSequence, values);

		if (length > 1) {
			if (!(info -> flags & UTFlagLowerExpands)) {
				fprintf (stderr, "Missing lower special seqeunce for '%04x'", value);
				return 1;
			}

			if (UTSpecialCases [info -> cases [UTCaseLower]] != length) {
				fprintf (stderr, "Wrong lower seqeunce length for '%04x'", value);
				return 1;
			}
		}

		length = getSequence (titleSequence, values);

		if (length > 1) {
			if (!(info -> flags & UTFlagTitleExpands)) {
				fprintf (stderr, "Missing title special seqeunce for '%04x'", value);
				return 1;
			}

			if (UTSpecialCases [info -> cases [UTCaseTitle]] != length) {
				fprintf (stderr, "Wrong title seqeunce length for '%04x'", value);
				return 1;
			}
		}
	}

	fclose (data);
	fclose (specialCasing);

	printf ("-- all good --\n");

	return 0;
}


test/test.h
#include <assert.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include "unicode-table.h"

#define RESULT_PASS   0
#define RESULT_FAIL   1
#define RESULT_SKIP  77
#define RESULT_ERROR 99


test/test-numbers.c
#include "test.h"

int main(int argc, char const* argv[])
{
	UTGlyph glyph;
	UTInfo const* info;

	// character `Ⅶ` (0x2166; ROMAN NUMERAL SEVEN)
	glyph = 0x2166;
	info = UTLookupGlyph(glyph);

	// check if character is a number
	if (info->flags & UT_FLAG_NUMBER) {
		// prints "Integer value of 0x2166: 7"
		printf("Integer value of 0x%04X: %lld\n", glyph, info->num);
	}
	else {
		return RESULT_FAIL;
	}

	// character `¼` (0x00BC; VULGAR FRACTION ONE QUARTER)
	glyph = 0x00BC;
	info = UTLookupGlyph(glyph);

	// check if character is a fraction
	if (info->flags & UT_FLAG_FRACTION) {
		// prints "String representation of 0x00BC: 1/4"
		printf("String representation of 0x%04X: %s\n", glyph, info->frac);
	}
	else {
		return RESULT_FAIL;
	}

	return RESULT_PASS;
}


src/generate.pl
#!/usr/bin/env perl

# Copyright (c) 2016-2017 Simon Schoenenberger
# https://github.com/detomon/unicode-table
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.

use strict;
use warnings;

use Template;

#-------------------------------------------------------------------------------
#
# Unicode definitions
#
#-------------------------------------------------------------------------------

use constant unicodeVersion => '15.0.0';
use constant tableSize      => 0x110000;

use constant moLetterGlyphInfo       => 1 << 0;
use constant moUppercaseGlyphInfo    => 1 << 1;
use constant moLowercaseGlyphInfo    => 1 << 2;
use constant moTitlecaseGlyphInfo    => 1 << 3;
use constant moSpaceGlyphInfo        => 1 << 4;
use constant moLinebreakGlyphInfo    => 1 << 5;
use constant moPunctuationGlyphInfo  => 1 << 6;
use constant moDigitGlyphInfo        => 1 << 7;
use constant moNumberGlyphInfo       => 1 << 8;
use constant moFractionGlyphInfo     => 1 << 9;
use constant moControlGlyphInfo      => 1 << 10;
use constant moSymbolGlyphInfo       => 1 << 11;
use constant moOtherGlyphInfo        => 1 << 12;
use constant moUpperExpandsGlyphInfo => 1 << 13;
use constant moLowerExpandsGlyphInfo => 1 << 14;
use constant moTitleExpandsGlyphInfo => 1 << 15;

my %categoryFlags = (
	''   => 0,
	'Lu' => moLetterGlyphInfo | moUppercaseGlyphInfo,
	'Ll' => moLetterGlyphInfo | moLowercaseGlyphInfo,
	'Lt' => moLetterGlyphInfo | moTitlecaseGlyphInfo,
	'Lm' => moLetterGlyphInfo,
	'Lo' => moLetterGlyphInfo,
	'Mn' => moOtherGlyphInfo,
	'Mc' => moOtherGlyphInfo,
	'Me' => moOtherGlyphInfo,
	'Nd' => moLetterGlyphInfo | moDigitGlyphInfo | moNumberGlyphInfo,
	'Nl' => moLetterGlyphInfo | moNumberGlyphInfo,
	'No' => moLetterGlyphInfo | moNumberGlyphInfo,
	'Pc' => moPunctuationGlyphInfo,
	'Pd' => moPunctuationGlyphInfo,
	'Ps' => moPunctuationGlyphInfo,
	'Pe' => moPunctuationGlyphInfo,
	'Pi' => moPunctuationGlyphInfo,
	'Pf' => moPunctuationGlyphInfo,
	'Po' => moPunctuationGlyphInfo,
	'Sm' => moSymbolGlyphInfo,
	'Sc' => moSymbolGlyphInfo,
	'Sk' => moSymbolGlyphInfo,
	'So' => moSymbolGlyphInfo,
	'Zs' => moSpaceGlyphInfo,
	'Zl' => moSpaceGlyphInfo | moLinebreakGlyphInfo,
	'Zp' => moSpaceGlyphInfo | moLinebreakGlyphInfo,
	'Cc' => moControlGlyphInfo,
	'Cf' => moOtherGlyphInfo,
	'Cs' => moOtherGlyphInfo,
	'Co' => moOtherGlyphInfo,
	'Cn' => moOtherGlyphInfo,
);

my %categoryIndexes = (
	''   => 0,
	'Lu' => 1,
	'Ll' => 2,
	'Lt' => 3,
	'Lm' => 4,
	'Lo' => 5,
	'Mn' => 6,
	'Mc' => 7,
	'Me' => 8,
	'Nd' => 9,
	'Nl' => 10,
	'No' => 11,
	'Pc' => 12,
	'Pd' => 13,
	'Ps' => 14,
	'Pe' => 15,
	'Pi' => 16,
	'Pf' => 17,
	'Po' => 18,
	'Sm' => 19,
	'Sc' => 20,
	'Sk' => 21,
	'So' => 22,
	'Zs' => 23,
	'Zl' => 24,
	'Zp' => 25,
	'Cc' => 26,
	'Cf' => 27,
	'Cs' => 28,
	'Co' => 29,
	'Cn' => 30,
);

my %categoryName = (
	'',  => 'CategoryInvalid',
	'Lu' => 'CategoryLetterUppercase',
	'Ll' => 'CategoryLetterLowercase',
	'Lt' => 'CategoryLetterTitlecase',
	'Lm' => 'CategoryLetterModifier',
	'Lo' => 'CategoryLetterOther',
	'Mn' => 'CategoryMarkNonspacing',
	'Mc' => 'CategoryMarkSpacingCombining',
	'Me' => 'CategoryMarkEnclosing',
	'Nd' => 'CategoryNumberDecimalDigit',
	'Nl' => 'CategoryNumberLetter',
	'No' => 'CategoryNumberOther',
	'Pc' => 'CategoryPunctuationConnector',
	'Pd' => 'CategoryPunctuationDash',
	'Ps' => 'CategoryPunctuationOpen',
	'Pe' => 'CategoryPunctuationClose',
	'Pi' => 'CategoryPunctuationInitialQuote',
	'Pf' => 'CategoryPunctuationFinalQuote',
	'Po' => 'CategoryPunctuationOther',
	'Sm' => 'CategorySymbolMath',
	'Sc' => 'CategorySymbolCurrency',
	'Sk' => 'CategorySymbolModifier',
	'So' => 'CategorySymbolOther',
	'Zs' => 'CategorySeparatorSpace',
	'Zl' => 'CategorySeparatorLine',
	'Zp' => 'CategorySeparatorParagraph',
	'Cc' => 'CategoryOtherControl',
	'Cf' => 'CategoryOtherFormat',
	'Cs' => 'CategoryOtherSurrogate',
	'Co' => 'CategoryOtherPrivateUse',
	'Cn' => 'CategoryOtherNotAssigned',
);

#-------------------------------------------------------------------------------
#
# Arguments
#
#-------------------------------------------------------------------------------

if (($#ARGV + 1) < 2) {
	print "usage $0 UnicodeData.txt SpecialCasing.txt\n";
	exit 1;
}

my $args = join ' ', @ARGV;
my $prefix = 'UT';
my $makeSnakeCase = 0;
my $useCategories = 0;
my %useCategories = ();
my $includeInfos = 'flags,categories,casing,numbers';
my %includeInfos = ();
my $excludeSurrogates = 0;
my %conditionalFlags = ();

if ($args =~ /--symbol-prefix=([\w_]+)/) {
	$prefix = $1;
}

if ($args =~ /--snake-case=(\d+)/) {
	$makeSnakeCase = int $1;
}

if ($args =~ /--categories=([\w_,]+)/) {
	foreach (split /,/, $1) {
		$useCategories {$_} = 1;
	}

	$useCategories = 1;
}

if ($args =~ /--include-info=([\w_,]+)/) {
	$includeInfos = $1;
}

if ($args =~ /--strict-level=(\d+)/) {
	$excludeSurrogates = int $1 > 0;
}

# full format: '{%1$5d,%2$3d,{%3$6d,%4$6d,%5$6d}, {%6$s}},'
my $infoFormat = '';
my %infoFormat = ();
my @infoFormat = ();

foreach (split /,/, $includeInfos) {
	$infoFormat {$_} = 1;
}

if (exists $infoFormat {'flags'}) {
	push @infoFormat, '%1$5d';
	$conditionalFlags {'addFlags'} = 1;
}

if (exists $infoFormat {'categories'}) {
	push @infoFormat, '%2$3d';
	$conditionalFlags {'addCategories'} = 1;
}

if (exists $infoFormat {'casing'}) {
	push @infoFormat, '{%3$6d,%4$6d,%5$6d}';
	$conditionalFlags {'addCasing'} = 1;
}

if (exists $infoFormat {'numbers'}) {
	push @infoFormat, ' {%6$s}';
	$conditionalFlags {'addNumbers'} = 1;
}

$infoFormat = (join ',', @infoFormat);
$infoFormat =~ s/^\s+|\s+$//g;
$infoFormat = "{$infoFormat},";

my $outName   = 'unicode-table';
my $hdrFile   = "$outName.h";
my $hdrFileIn = "unicode-table.h.in";
my $srcFile   = "$outName.c";
my $srcFileIn = "unicode-table.c.in";

#-------------------------------------------------------------------------------
#
# Prepare tables
#
#-------------------------------------------------------------------------------

my @data           = (0) x tableSize;
my %special        = ();
my %types          = (sprintf ($infoFormat, 0, 0, 0, 0, 0, 0) => 0);
my @pages          = (0) x (tableSize >> 8);
my %pageCache      = ();
my @specialCasing  = ();

$pageCache {join ',', ((0) x 256)} = 0;

#-------------------------------------------------------------------------------
#
# Functions
#
#-------------------------------------------------------------------------------

sub makeCharSequence {
	my $codes = shift;

	$codes =~/\s*(.+)\s*/;

	my @sequence = split /\s+/, $1;
	my $offset   = $#specialCasing + 1;

	# ignore single glyph
	return -1 if ($#sequence == 0);

	push @specialCasing, $#sequence + 1;
	push @specialCasing, hex ($_) foreach (@sequence);

	return $offset;
}

sub getTypeIndex {
	my $info = shift;
	my $catIdx = shift;
	my $upper = shift;
	my $lower = shift;
	my $title = shift;
	my $number = shift;

	my $type = sprintf ($infoFormat, $info, $catIdx, $upper, $lower, $title, $number);

	if ($types {$type}) {
		$type = $types {$type};
	}
	else {
		my $count = keys %types;

		$types {$type} = $count;
		$type = $count;
	}

	return $type;
}

sub unsignedTypeFromSize {
	my $size = shift;
	my $type = 'uint32_t';

	if ($size <= 0xFF) {
		$type = 'uint8_t';
	}
	elsif ($size <= 0xFFFF) {
		$type = 'uint16_t';
	}

	return $type;
}

#-------------------------------------------------------------------------------
#
# Read special cases
#
#-------------------------------------------------------------------------------

open my $specialFile, '<', $ARGV[1] or die "File '$ARGV[1]' not found";

while (<$specialFile>) {
	chomp;

	# ignore empty lines and comments
	next if ($_ =~ /^$|^#/);

	$_ =~ /(.+);\s*#/;

	my @line = split ';', $1;

	# ignore conditional case-folding
	next if ($line [4]);

	my $code  = hex ($line [0]);
	my $lower = makeCharSequence $line [1];
	my $title = makeCharSequence $line [2];
	my $upper = makeCharSequence $line [3];

	my @cases = ($upper, $lower, $title);

	@{$special {$code}} = @cases;

	@line = split ';', $_;
}

close $specialFile;

#-------------------------------------------------------------------------------
#
# Read unicode data
#
#-------------------------------------------------------------------------------

my %specialChars = (
	0x0009 => moSpaceGlyphInfo,                        # CHARACTER TABULATION
	0x000A => moSpaceGlyphInfo | moLinebreakGlyphInfo, # LINE FEED (LF)
	0x000B => moSpaceGlyphInfo,                        # LINE TABULATION
	0x000C => moSpaceGlyphInfo,                        # FORM FEED (FF)
	0x000D => moSpaceGlyphInfo | moLinebreakGlyphInfo, # CARRIAGE RETURN (CR)
	0xFEFF => moSpaceGlyphInfo,                        # ZERO WIDTH NO-BREAK SPACE (BYTE ORDER MARK)
);

open my $dataFile, '<', $ARGV[0] or die "File '$ARGV[0]' not found";

while (<$dataFile>) {
	chomp;

	my @line = split /;/, $_;
	my $code = hex ($line [0]);
	my $cat  = $line [2];
	my $info = $categoryFlags {$cat};

	my $number = $line [8];
	my $upper  = hex ($line [12] || 0);
	my $lower  = hex ($line [13] || 0);
	my $title  = hex ($line [14] || 0);

	if ($useCategories && !$useCategories {$cat}) {
		$info   = moOtherGlyphInfo;
		$cat    = 'Cn';
		$number = 0;
		$upper  = 0;
		$lower  = 0;
		$title  = 0;
	}
	else {
		if ($specialChars {$code}) {
			$info |= $specialChars {$code};
		}

		$upper = $upper - $code if ($upper);
		$lower = $lower - $code if ($lower);
		$title = $title - $code if ($title);

		if ($number =~ /\//) {
			my ($v1, $v2) = split '/', $number;

			$number = ".frac=\"$v1/$v2\"";
			$info |= moFractionGlyphInfo;
		}
		elsif ($info & moNumberGlyphInfo) {
			$number = int ($number);
		}
		else {
			$number = 0;
		}

		if ($special {$code}) {
			my @cases = @{$special {$code}};

			if ($cases [0] != -1) {
				$upper = $cases [0];
				$info |= moUpperExpandsGlyphInfo;
			}

			if ($cases [1] != -1) {
				$lower = $cases [1];
				$info |= moLowerExpandsGlyphInfo;
			}

			if ($cases [2] != -1) {
				$title = $cases [2];
				$info |= moTitleExpandsGlyphInfo;
			}
		}
	}

	my $type = getTypeIndex ($info, $categoryIndexes {$cat}, $upper, $lower, $title, $number);

	# read range
	if ($line [1] =~ /First>$/i) {
		$_ = <$dataFile>;
		chomp;

		my @line2 = split /;/, $_;
		my $code2 = hex ($line2 [0]);

		if ($excludeSurrogates) {
			if ($cat =~ /Cs/i) {
				next;
			}
		}

		for (; $code <= $code2; $code ++) {
			$pages [$code >> 8] = 1;
			$data [$code] = $type;
		}
	}
	else {
		$pages [$code >> 8] = 1;
		$data [$code] = $type;
	}
}

close $dataFile;

#-------------------------------------------------------------------------------
#
# Build pages cache
#
#-------------------------------------------------------------------------------

my $cacheCount = 1;

for (my $i = 0; $i <= $#pages; $i ++) {
	next unless ($pages [$i]);

	my $index = 0;
	my $page  = join ',',  @data [($i << 8) .. ((($i + 1) << 8) - 1)];

	if ($pageCache {$page}) {
		$index = $pageCache {$page};
	}
	else {
		$index = $cacheCount ++;
		$pageCache {$page} = $index;
	}

	$pages [$i] = $index;
}

#-------------------------------------------------------------------------------
#
# Print
#
#-------------------------------------------------------------------------------

my @infoKeys  = keys %types;
my $infoSize  = @infoKeys;
my $pagesSize = keys %pageCache;
my $infoType  = unsignedTypeFromSize $infoSize;
my $pagesType = unsignedTypeFromSize $pagesSize;

my %printMethods = ();

my $template = new Template(
	'vars' => {
		'outName'   => $outName,
		'infoType'  => $infoType,
		'pagesType' => $pagesType,
	},
	'prefix' => $prefix,
	'makeSnakeCase' => $makeSnakeCase,
	'printMethods' => \%printMethods,
	'conditional' => sub {
		return exists $conditionalFlags {$_[0]};
	},
);

%printMethods = (
	'header' => sub {
		my $out = shift;

		print $out " * Generated by $0\n";
		print $out " * Unicode version ".unicodeVersion."\n";
		print $out " * https://github.com/detomon/unicode-table\n";
	},
	'categories' => sub {
		my $out = shift;

		foreach (sort { $categoryIndexes {$a} <=> $categoryIndexes {$b} } keys %categoryIndexes) {
			my $line = $categoryName {$_};

			$line = $template->toConstant($line);
			$line = sprintf "\t%-39s ///< %s", "$line,", $_;
			$line =~ s/\s+$//;

			print $out "$line\n";
		}
	},
	'infos' => sub {
		my $out = shift;

		foreach (sort { $types {$a} <=> $types {$b} } keys %types) {
			print $out "	$_\n";
		}
	},
	'pageIndex' => sub {
		my $out = shift;
		my $i = 0;
		my $p = 0;

		print $out "\t";

		foreach (@pages) {
			print $out "\n\t" if ($i > 0 && $i % 16 == 0);
			printf $out "%3d,", $_;

			$i ++;
		}

		print $out "\n";
	},
	'infoIndex' => sub {
		my $out = shift;
		my $p = 0;
		my $i = 0;

		foreach (sort { $pageCache {$a} <=> $pageCache {$b} } keys %pageCache) {
			print $out "\t{" if ($p == 0);
			print $out "\n\t}, {" if ($p > 0);

			foreach (split /,/, $_) {
				print $out "\n\t" if ($i % 16 == 0);
				printf $out "%3d,", $_;

				$i ++;
			}

			$p ++;
		}

		print $out "\n\t}\n";
	},
	'specialCases' => sub {
		my $out = shift;
		my $line = '';

		foreach (@specialCasing) {
			my $data = sprintf "%d, ", $_;

			if (length ($line) + length ($data) >= 66) {
				$line =~ s/\s+$//;
				print $out "\t$line\n";

				$line = $data;
			}
			else {
				$line .= $data;
			}
		}

		$line =~ s/\s+$//;
		print $out "\t$line\n";
	},
	'categoryNames' => sub {
		my $out = shift;

		foreach (sort { $categoryIndexes {$a} <=> $categoryIndexes {$b} } keys %categoryIndexes) {
			my $key = $categoryName {$_};

			$key = $template->toConstant($key);

			printf $out "\t%-39s = \"%s\",\n", "[$key]", $_;
		}
	},
);

sub main {
	open my $hdrin,  '<', $hdrFileIn or die "File '$hdrFileIn' not found";
	open my $hdrout, '>', $hdrFile or die "File '$hdrFile' not writable";

	open my $srcin,  '<', $srcFileIn or die "File '$srcFileIn' not found";
	open my $srcout, '>', $srcFile or die "File '$srcFile' not writable";

	$template->readLines($hdrin, $hdrout);
	$template->readLines($srcin, $srcout);

	close $hdrin;
	close $hdrout;

	close $srcin;
	close $srcout;

	return 0;
}

exit main @ARGV;
